### PR TITLE
Keep shebang directives intact

### LINKED ISSUE #2367: Shebang directives get removed from scripts
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKBiAhAegK4GcAnbAIwEsA7bRCgNzAFoBlMeWAFwsXbADN8yjBgHcAhoQooArELETYkgIyKAHACZZ4iqPyJChWIQDUAHUImKZi7AAONMEwCe%2BdogC2VqtjAA6P58gAGnAoXXZ2SgBzfFQwAG0QiFAIFKgAEnZHO1jISnYgxJTINNpRABtcZDQE1NSABmDaooBJCiQKdiYyAC9kRqawABZCiABdQoBffqSR9MzstFyOgqbi0oqq%2BNmIRWnayABZUQAPABlKRFOaSPYACxWBxTU62fHUqcLk-YyszcgaeAAeV45y4TEyZT6szW5UqsRqAzUe1SkAAom1gaCoQMoGVeJBXpM9l8UT8FlASLBYGUHqT1nDqtswABmZEtCi6QjsABilHKADlEMIyhdabV2IRKoT3sToWS-pTqWKIDCNvCmYM2SqmDZRNBEAAhRC8QyIAAK4lEbm4emVKQlUqabxSH1SJKK8pyippbNVDK2OKkWqgOr1huNptOsGEemgOkQrVosDjEQUdogDsQ0pdstWnsW3rtfs2CKaADZg5BQ-qjSbCIgAKo2OyEOO6RPJ0SpijpvjlXTZiCulLulX5ilUn1y%2BklpkAdkr1fDdcQAGEyjp8KuFC5JdB2IZe7x%2B1mnUTPnL5grJ0WSrDZziVIvdTWI-WDu4SLbg8eygOzzKF55leXo3r6d5qoyOIAJzPmGtamtuZRpj%2BJ6DmAw4zMBvygUq4EzuqOKKA0TJVi%2BiAAIK8K4hDbm4biiL2mboZhYCjnMOEFmB073oRjy7KRS4IfWTDuGQcDIT2qF-qetTOkOubfCBXF4TxkEBo8SKCeRVE0aJbjidSKFMsxAE5kBSmcROql5gRUGPKy2lhhRBi4G0AAiiAigZNFMZKsmpPJGGKaSylQHkt52RpTSKJqpFHMczS8AAKrcNBMLchjsAA6mQ8B3L2LxmQpFmhVZSz5PhvH2TFQbxScSWpTQqIyRlWW5fl9zBoMRVyeebqXuVEVVeppa1IoFb1YlFC8GQxyAi2XaGKixw2PW%2BACMZga9YF-UjoN5IVZF1XReNC5TQASogcCEPAHUFd1O0pEFrHscUYVHSN-pjakqiVglV03fA-K4G4X6EMCzSuG4MTBooLEhR6H1uLgZQRCKXDcoYDHhHoyU4V9D6PLBpGA4Y8AHKj6MXFjhA475lbQLclr7noAD6wh5QVCOlUjQ3LITfFNM8-0nC5hCiI4gKEOcLj3V1GpPWMe1YZZh3DWp31MmoAk4ocYv6JL0uy%2BwINg3okPQ7DTLw8VwW82OyNU2QGOILT9N4wTmtE8LWl6%2BLRsy2QLiU2jLs09jXYM6RTMszRHNc-cPMDdh6sC97Qu1Gojl6wlABq94GpQ8BRPLvYqErYAvYjjv85VGc1VncW5yc3JufuZAKEXbSl4nvY9cn%2B2p38Gu2SdP0pGodUt8c7kcAA4twK1rYgG2dxQZfBlIlfVw7HGHYWguN6kaiTbnzuuwayHQAA1gaEu39w%2BCAhQTBWmu1Kg1JTK-v%2BfWASnNW14bLfCihPCAahzp60FMKC4Rp2AxhoPjOwnlZoUDIN2CibQPzm0INbHEv8ArPRVmxA6wCpxj1GtrJ8pEKIikiBQNuFAO4KC6AwrsuB6zJVgK0do7AuzryPGhO2r0yG4QoaA8e2sSb%2BzRnobQERaCXAUJEHB4NUGUAwevfBAxCGD1VmVA%2B3FKFaxxMyEi58w5YlOFaEg8BRDrlgAICgkQYGuyETJfRpDh7iOOlQsxusBholWnoMg1oOjlAANKIEQDYXhNB2CtHvqIZhCsCHCP-uZQBhjyF%2BNMQMZkfsglXxSTfLE%2BAXKwDcsDVergKYXwuMvdam0OR%2BUdJkkq2S%2BZGJAXSKRZic7FPEMJRA7lg7QEIGEvkdSGzoIUJ5aAG4Jbdg8X-XaACh5AN8UfU6qRmTNyCcvUJ4T%2BFlHBK5XcuAbDgkcJCVZRDlYbIMd0ke6cTE%2B1qMyaeQTomxISmbcGwISkUDKRcHRMU6g7xIW9cckBD4N12SkZkZ8gnnPEuwA4CBsS6Iyesl0IRRj9FyPgbkAhYh6ImEAA)

#### Code

```fsharp
#!/usr/bin/env -S dotnet fsi --warn:5 --warnon:1182 --warnaserror+

open System
// ...

```
            

#### Result

```fsharp
open System
// ...

```
            
#### Problem description

See fsharp/fslang-design#683.

#### Extra information

- [x] The formatted result breaks my code.
- [ ] The formatted result gives compiler warnings.
- [x] I or my company would be willing to help fix this.

#### Options

Fantomas master branch at 2022-07-15T14:30:45Z - 87cd2120c82ce2c83838ee0eaa81f53d03ed33af

```fsharp
    { config with
                MaxIfThenElseShortWidth = 40
                MaxInfixOperatorExpression = 50
                MaxArrayOrListWidth = 40
                MaxDotGetExpressionWidth = 50
                NewlineBetweenTypeDefinitionAndMembers = false }
```

<sub>Did you know that you can ignore files when formatting from fantomas-tool or the FAKE targets by using a [.fantomasignore file](https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md#ignore-files-fantomasignore)?</sub>

### PR DESCRIPTION
**Pull request by mataha**
Fixes #2367.

### CODE REVIEW

File: src/Fantomas.Core.Tests/ShebangTests.fs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    let codeSnippet =
+        """#!/usr/bin/env -S dotnet fsi
+// random licensing stuff
+
+open System
+
+printfn "the best thing we've ever done"
+"""
+
+    formatSourceString false codeSnippet config |> should equal codeSnippet
```
Comment (nojaf) [inline src/Fantomas.Core.Tests/ShebangTests.fs:18]: Please use the typical test template, https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#pull-request-ground-rules.

Even if the before and after are the same, all our (2000+) other tests look like this. There is no need to diverge from this.

File: src/Fantomas.Core.Tests/ShebangTests.fs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+[<Test>]
```
Comment (nojaf) [inline src/Fantomas.Core.Tests/ShebangTests.fs:20]: Please remove these validation tests, you are testing the parser more than the actual Fantomas code.
`formatSourceString` will also already validate the code.

https://github.com/fsprojects/fantomas/blob/c81e47465df56e4b49e6aab63f25ea9adf56ed90/src/Fantomas.Core.Tests/TestHelpers.fs#L23-L35

File: src/Fantomas.Core/Trivia.fs [hunk 3]
```diff
@@ -80,11 +80,17 @@ let internal collectTriviaFromCodeComments
         | CommentTrivia.LineComment r ->
             let content = source.GetContentAt r
 
-            let line = source.GetLineString(r.StartLine - 1)
+            let index = r.StartLine - 1
+
```
Comment (nojaf) [inline src/Fantomas.Core/Trivia.fs:84]: Could you put these all after each other maybe?
```fsharp
let content = ...
let index = ...
let line = ...
```

**Review by nojaf — CHANGES_REQUESTED** [review]
Thank you @mataha for this first contribution.

File: src/Fantomas.Core.Tests/ShebangTests.fs (hunk 1 above)
Comment (mataha) [reply src/Fantomas.Core.Tests/ShebangTests.fs]: Done!

File: src/Fantomas.Core.Tests/ShebangTests.fs (hunk 2 above)
Comment (mataha) [reply src/Fantomas.Core.Tests/ShebangTests.fs]: Done!

File: src/Fantomas.Core/Trivia.fs (hunk 3 above)
Comment (mataha) [reply src/Fantomas.Core/Trivia.fs]: Done!

**Review by nojaf — APPROVED** [review]
Thank you @mataha!
Would you like to have a (beta) release with this fix?

**Comment by mataha** [comment]
> Thank you @mataha! Would you like to have a (beta) release with this fix?

If that's not a burden - sure.

**Pull request merged**